### PR TITLE
Add functional methods for instances, Instance class and tests

### PR DESCRIPTION
**Pull request by mziccard**
This PR adds functional methods for instance and functional `Instance` class.
This PR also adds unit tests for `ComputeImpl` and `Instance` as well as integration tests.

### CODE REVIEW

File: gcloud-java-compute/src/test/java/com/google/gcloud/compute/it/ITComputeTest.java [hunk 1]
```diff
@@ -23,11 +23,19 @@
...
 import com.google.common.collect.ImmutableSet;
 import com.google.gcloud.Page;
 import com.google.gcloud.compute.Address;
 import com.google.gcloud.compute.AddressId;
 import com.google.gcloud.compute.AddressInfo;
+import com.google.gcloud.compute.AttachedDisk;
+import com.google.gcloud.compute.AttachedDisk.AttachedDiskConfiguration;
+import com.google.gcloud.compute.AttachedDisk.CreateDiskConfiguration;
+import com.google.gcloud.compute.AttachedDisk.PersistentDiskConfiguration;
+import com.google.gcloud.compute.AttachedDisk.PersistentDiskConfiguration.Mode;
```
Comment (mziccard) [inline gcloud-java-compute/src/test/java/com/google/gcloud/compute/it/ITComputeTest.java]: Codacy Issue found: [Avoid unused imports such as 'com.google.gcloud.compute.AttachedDisk.PersistentDiskConfiguration.Mode'](https://www.codacy.com/app/mziccard/gcloud-java/file/2464490828/issues/source?bid=3140968&fileBranchId=3256314#l37)

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java [hunk 2]
```diff
@@ -1046,6 +1093,54 @@ public static NetworkFilter notEquals(NetworkField field, boolean value) {
...
+    }
+
+    /**
+     * Returns an equals filter for the given field and string value. For string fields,
+     * {@code value} is interpreted as a regular expression using RE2 syntax. {@code value} must
+     * match the entire field.
+     *
+     * @see <a href="https://github.com/google/re2/wiki/Syntax">RE2</a>
+     */
+    public static InstanceFilter equals(InstanceField field, String value) {
```
Comment (mziccard) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java]: Codacy Issue found: [The method name and parameter number are suspiciously close to equals(Object)](https://www.codacy.com/app/mziccard/gcloud-java/file/2464490711/issues/source?bid=3140968&fileBranchId=3256314#l1113)

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java [hunk 3]
```diff
@@ -1046,6 +1093,54 @@ public static NetworkFilter notEquals(NetworkField field, boolean value) {
...
+     * @see <a href="https://github.com/google/re2/wiki/Syntax">RE2</a>
+     */
+    public static InstanceFilter notEquals(InstanceField field, String value) {
+      return new InstanceFilter(checkNotNull(field), ComparisonOperator.NE, checkNotNull(value));
+    }
+
+    /**
+     * Returns a equals filter for the given field and boolean value.
+     */
+    public static InstanceFilter equals(InstanceField field, boolean value) {
```
Comment (mziccard) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java]: Codacy Issue found: [The method name and parameter number are suspiciously close to equals(Object)](https://www.codacy.com/app/mziccard/gcloud-java/file/2464490711/issues/source?bid=3140968&fileBranchId=3256314#l1131)

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java [hunk 4]
```diff
@@ -1606,6 +1648,399 @@ public Operation deleteNetwork(String network, OperationOption... options) {
...
+      }
+    };
+  }
+
+  @Override
+  public Page<Instance> listInstances(String zone, InstanceListOption... options) {
+    return listInstances(zone, options(), optionMap(options));
+  }
+
+  private static Page<Instance> listInstances(final String zone,
```
Comment (mziccard) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java]: Codacy Issue found: [Avoid unused private methods such as 'listInstances(String,ComputeOptions,Map)'.](https://www.codacy.com/app/mziccard/gcloud-java/file/2464490714/issues/source?bid=3140968&fileBranchId=3256314#l1702)

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java [hunk 5]
```diff
@@ -1606,6 +1648,399 @@ public Operation deleteNetwork(String network, OperationOption... options) {
...
+      throw ComputeException.translateAndThrow(e);
+    }
+  }
+
+  @Override
+  public Page<Instance> listInstances(InstanceAggregatedListOption... options) {
+    return listInstances(options(), optionMap(options));
+  }
+
+  private static Page<Instance> listInstances(final ComputeOptions serviceOptions,
```
Comment (mziccard) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java]: Codacy Issue found: [Avoid unused private methods such as 'listInstances(ComputeOptions,Map)'.](https://www.codacy.com/app/mziccard/gcloud-java/file/2464490714/issues/source?bid=3140968&fileBranchId=3256314#l1730)

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java (hunk 2 above)
Comment (mziccard) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java]: Codacy Issue found: [The method name and parameter number are suspiciously close to equals(Object)](https://www.codacy.com/app/mziccard/gcloud-java/file/2464634543/issues/source?bid=3140968&fileBranchId=3256314#l1113)

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java (hunk 3 above)
Comment (mziccard) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java]: Codacy Issue found: [The method name and parameter number are suspiciously close to equals(Object)](https://www.codacy.com/app/mziccard/gcloud-java/file/2464634543/issues/source?bid=3140968&fileBranchId=3256314#l1131)

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java [hunk 6]
```diff
@@ -1605,6 +1647,399 @@ public Operation deleteNetwork(String network, OperationOption... options) {
...
+      }
+    };
+  }
+
+  @Override
+  public Page<Instance> listInstances(String zone, InstanceListOption... options) {
+    return listInstances(zone, options(), optionMap(options));
+  }
+
+  private static Page<Instance> listInstances(final String zone,
```
Comment (mziccard) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java]: Codacy Issue found: [Avoid unused private methods such as 'listInstances(String,ComputeOptions,Map)'.](https://www.codacy.com/app/mziccard/gcloud-java/file/2464634546/issues/source?bid=3140968&fileBranchId=3256314#l1701)

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java [hunk 7]
```diff
@@ -1605,6 +1647,399 @@ public Operation deleteNetwork(String network, OperationOption... options) {
...
+      throw ComputeException.translateAndThrow(e);
+    }
+  }
+
+  @Override
+  public Page<Instance> listInstances(InstanceAggregatedListOption... options) {
+    return listInstances(options(), optionMap(options));
+  }
+
+  private static Page<Instance> listInstances(final ComputeOptions serviceOptions,
```
Comment (mziccard) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java]: Codacy Issue found: [Avoid unused private methods such as 'listInstances(ComputeOptions,Map)'.](https://www.codacy.com/app/mziccard/gcloud-java/file/2464634546/issues/source?bid=3140968&fileBranchId=3256314#l1729)

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java [hunk 8]
```diff
@@ -1605,6 +1647,399 @@ public Operation deleteNetwork(String network, OperationOption... options) {
...
+            }
+          }, options().retryParams(), EXCEPTION_HANDLER);
+      return answer == null ? null : Instance.fromPb(this, answer);
+    } catch (RetryHelper.RetryHelperException e) {
+      throw ComputeException.translateAndThrow(e);
+    }
+  }
+
+  private static Function<com.google.api.services.compute.model.Instance, Instance>
+      instanceFromPb(final ComputeOptions serviceOptions) {
```
Comment (jgeewax) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java]: What's the difference between this and Instace.fromPb above?

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/spi/DefaultComputeRpc.java [hunk 9]
```diff
@@ -869,6 +880,256 @@ public Operation deleteNetwork(String network, Map<Option, ?> options) {
...
+      return nullForNotFound(ex);
+    }
+  }
+
+  @Override
+  public Tuple<String, Iterable<Instance>> listInstances(String zone, Map<Option, ?> options) {
+    try {
+      InstanceList instanceList = compute.instances()
+          .list(this.options.projectId(), zone)
+          .setFilter(FILTER.getString(options))
```
Comment (jgeewax) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/spi/DefaultComputeRpc.java]: this pattern seems backwards to me (saying "key".getfrom(map), right?), but OK.

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/spi/DefaultComputeRpc.java [hunk 10]
```diff
@@ -869,6 +880,256 @@ public Operation deleteNetwork(String network, Map<Option, ?> options) {
...
+      throw translate(ex);
+    }
+  }
+
+  @Override
+  public Operation deleteInstance(String zone, String instance, Map<Option, ?> options) {
+    try {
+      return compute.instances()
+          .delete(this.options.projectId(), zone, instance)
+          .setFields(FIELDS.getString(options))
```
Comment (jgeewax) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/spi/DefaultComputeRpc.java]: Why do we need fields when deleting an instance?

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/spi/DefaultComputeRpc.java [hunk 11]
```diff
@@ -869,6 +880,256 @@ public Operation deleteNetwork(String network, Map<Option, ?> options) {
...
+  }
+
+  @Override
+  public Operation deleteAccessConfig(String zone, String instance, String networkInterface,
+      String accessConfig, Map<Option, ?> options) {
+    try {
+      return compute.instances()
+          .deleteAccessConfig(this.options.projectId(), zone, instance, accessConfig,
+              networkInterface)
+          .setFields(FIELDS.getString(options))
```
Comment (jgeewax) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/spi/DefaultComputeRpc.java]: Ditto to above?

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java (hunk 8 above)
Comment (mziccard) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java]: This is a factory method for conversion functions. You pass the service options and this returns a `Function` object that does exactly what `Instance.fromPb` does. We need this to create `Function` objects for conversion to pass to Guava's `Iterables.transform`.

File: gcloud-java-compute/src/test/java/com/google/gcloud/compute/it/ITComputeTest.java [hunk 12]
```diff
@@ -1599,4 +1616,307 @@ public void testAggregatedListSubnetworks() throws InterruptedException {
...
+    NetworkInterface networkInterface = NetworkInterface.builder(networkId).build();
+    AttachedDisk disk = AttachedDisk.of("dev0",
+        CreateDiskConfiguration.builder(IMAGE_ID).autoDelete(true).build());
+    InstanceInfo instanceInfo =
+        InstanceInfo.builder(instanceId, MachineTypeId.of(ZONE, MACHINE_TYPE))
+            .attachedDisks(disk)
+            .networkInterfaces(networkInterface)
+            .build();
+    Operation operation = compute.create(instanceInfo);
+    while (!operation.isDone()) {
```
Comment (jgeewax) [inline gcloud-java-compute/src/test/java/com/google/gcloud/compute/it/ITComputeTest.java]: we don't have a way to wait on operations?

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/spi/DefaultComputeRpc.java (hunk 10 above)
Comment (mziccard) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/spi/DefaultComputeRpc.java]: Delete returns an `Operation` resource and we can select which fields of such operation resource we want to return, for instance you might just be interested in the operation's id.

File: gcloud-java-compute/src/test/java/com/google/gcloud/compute/it/ITComputeTest.java (hunk 12 above)
Comment (mziccard) [inline gcloud-java-compute/src/test/java/com/google/gcloud/compute/it/ITComputeTest.java]: No we don't (and also for BigQuery jobs, that have a very similar pattern). The reason is that having an `operation.waitTermination()` method that simply hangs waiting for operation to be done seemed a bit odd to me. An alternative would be:
```java
operation.whenDone(new OperationCallback() {
  public void onSuccess(Operation operation) {
    // operation succeeded, operation.errors() is null
  }

  public void onFailure(List<OperationError> errors) {
    // handle operation failure
  }
});
```

The only concern I have with this latter approach is that it gives the idea of being asynchronous while in reality it blocks until the operation is done. Any opinion?

I will open an issue for this this to we can address it also for BigQuery and do not block this PR.

**Pull request merged**